### ISSUE TITLE
Auto-indent issues - square brackets, function arguments, heredoc/nowdoc

### ISSUE THREAD

**Issue by crantok** [atom/language-php#63]
Is this the right place to ask about indenting issues that I am experiencing when editing PHP? Or is auto indenting actually handled by a different module?

The first two examples have very obvious work-arounds and are just annoying. The heredoc issue requires any heredoc strings to be revisited every time anyone auto-indents the whole file.

Indenting square-bracket array syntax over multiple lines only works with a trailing coma, so this works:

```php
class Foo
{
	function bar ()
	{
		return
		[
			[
				// stuff
			],
			[
				// more stuff
			],
		];
	}
}
```

But this does not:

```php

class Foo
{
	function bar ()
	{
		return
		[
			[
				// stuff
			],
			[
				// more stuff
				]
			];
		}
	}	
```

Indenting  a multi-line function call only works if the closing parenthesis is on the same line as the final parameter, so this works:

```php
class Foo
{
	function bar ()
	{
		foobar(
		extremely_long_and_tedious_variable_name_1,
		extremely_long_and_tedious_variable_name_2,
		extremely_long_and_tedious_variable_name_3 );
	}
}
```

But this does not:

```php
class Foo
{
	function bar ()
	{
		foobar(
		extremely_long_and_tedious_variable_name_1,
		extremely_long_and_tedious_variable_name_2,
		extremely_long_and_tedious_variable_name_3
	);
}
}
```

And finally, heredoc and nowdoc syntax are both broken by auto-indent:

```php
class Foo
{
	function bar ()
	{
		return
		<<<'HTML'
		<blockquote>Somthing pithy</blockquote>
		HTML
	}
}
```

**Comment by JesseLeite** [comment]
@crantok Wonder if your issue is related to this PR? https://github.com/atom/language-php/pull/96

**Comment by crantok** [comment]
@JesseLeite Yes, that definitely relates to at least one of the cases above.

I stopped using atom in favour of lighttable soon after reporting this issue. LT uses CodeMirror for syntax highlighting and indenting. It is very easy to sanity-check issues using the [codemirror online demos](https://codemirror.net/mode/) and all formatting issues are concentrated in one project. I'm not saying that's the reason I swapped, just that it seems to provide a highly favourable workflow for dealing with formatting issues.

**Comment by charlesqwu** [comment]
Issue #63 (https://github.com/atom/language-php/issues/63) has two parts: 1. auto-indent issue related to () and [] -- that commit #141 fixes this part very well. I have tested many conditions related to this part, and it works fine. 2. auto-indent issue related to heredoc and newdoc.  The commit #141 does not fix this second part.

However, most PHP developers would experience the second issue a lot less frequently than the first issue, and so I would suggest that we merge the commit #141 as soon as possible.
